AccountOptions - fetch & store Api  created with documentation

engrshishir opened the pull request.

engrshishir committed:
Accounts Option Model & Migration created

engrshishir committed:
Modify 2024_03_30_200932_create_accounts_options_table

engrshishir committed:
AccountOptionService & AccountOptionRepository created

engrshishir committed:
AccountOption Store Api for backend

engrshishir committed:
API documentation for AccountOption Store backend route

engrshishir committed:
Create AccountOption seeder

engrshishir committed:
AccountOption fetch Api for backend

engrshishir committed:
API documentation for fetch AccountOptions backend route

engrshishir committed:
AccountOptions store Api documentation modified

engrshishir committed:
Backend AccountOption route is modified

engrshishir committed:
AccountOption fetch API for frontend

engrshishir committed:
AccountOptions store Api backend documentation modified

ManiruzzamanAkash commented on routes/api/v1/backend.php:
```diff
@@ -1,7 +1,11 @@
+Route::apiResource('donations', DonationsController::class)->only(['index', 'update', 'destroy']);
+
+// AccountOption (Bikash, Nagod, Rocket)
+Route::apiResource('accountoptions', AccountOptionController::class)->only(['index', 'store', 'update', 'destroy']);
```
Shoud be URL friendly route path - `account-options`

ManiruzzamanAkash commented on routes/api/v1/frontend.php:
```diff
@@ -11,3 +12,4 @@
 
 Route::get('/ping', [PingPongController::class, 'index']);
 Route::apiResource('donations', DonationsController::class)->only(['index', 'store']);
+Route::apiResource('accountoptions', AccountOptionController::class)->only(['index']);
```
`account-options`

ManiruzzamanAkash commented on database/seeders/AccountOptionSeeder.php:
```diff
@@ -0,0 +1,23 @@
+{
+    public function __construct(private readonly AccountOptionService $accountOptionService)
+    {
+    }
+    /**
+     * Run the database seeds.
```
Add extra space line after the constructor
```php
    public function __construct(private readonly AccountOptionService $accountOptionService)
    {
    }

    /**
     * Run the database seeds.
```

ManiruzzamanAkash commented on app/Http/Controllers/Backend/AccountOptionController.php:
```diff
@@ -0,0 +1,84 @@
+
+    /**
+     * @OA\GET(
+     *     path="/api/v1/backend/AccountOptions",
```
path should be always url friendly, instead of this -
`AccountOptions` to `account-options`

ManiruzzamanAkash replied:
Please change same in other places and definitely after change make a quick test must.

ManiruzzamanAkash commented on app/Http/Controllers/Backend/AccountOptionController.php:
```diff
@@ -0,0 +1,84 @@
+            // Return a JSON response with the retrieved account options and a success message
+            return $this->responseJson($options, Response::HTTP_OK, $message);
+        } catch (Exception $exception) {
+            // Handle any exceptions and return an error response
```
Remove the comment.

ManiruzzamanAkash commented on app/Http/Controllers/Backend/AccountOptionController.php:
```diff
@@ -0,0 +1,84 @@
+    public function index(): JsonResponse
+    {
+        try {
+            // Retrieve all account options
+            $options = $this->accountOptionService->getAccountOptions();
```
Some comments are not useful in my opinion.

We should do less comment, but good variable name, good coding. So that codes can be understand by reading without commnet. If needs comment, then we'll create a small function for that rather commenting.

So, just this is enough - 
```php
$accountOptions = $this->accountOptionService->getAccountOptions();

$totalMessage = 'Total ' . count($options) . ' ' . Str::plural('options', count($options)) . ' found.';

return $this->responseJson($accountOptions, Response::HTTP_OK, $totalMessage);
```

ManiruzzamanAkash commented on app/Http/Controllers/Backend/AccountOptionController.php:
```diff
@@ -0,0 +1,84 @@
+            return $this->responseErrorJson($exception);
+        }
+    }
+
+
+
```
Remove 2 extra space line, keep only one extra space after every method.

ManiruzzamanAkash commented on app/Http/Controllers/Frontend/AccountOptionController.php:
```diff
@@ -0,0 +1,50 @@
+    public function index(): JsonResponse
+    {
+        try {
+            // Retrieve all account options
```
Similar thing is applicable here like above for comments.

ManiruzzamanAkash commented on app/Http/Requests/AccountOptionRequest.php:
```diff
@@ -0,0 +1,28 @@
+    public function rules(): array
+    {
+        return [
+            'donation_option	' => ['string', 'max:150'],
```
This would not work maybe. There is extra space here
'donation_option ' would be 'donation_option'

ManiruzzamanAkash commented on app/Http/Requests/AccountOptionRequest.php:
```diff
@@ -0,0 +1,28 @@
+
+use Illuminate\Foundation\Http\FormRequest;
+
+class AccountOptionRequest extends FormRequest
```
We should not extend the default `FormRequest`cof Laravel, You can see, we've our own `FormRequest` by modifying the error response in a fashionable way, we need to use that.

Extend this `FormRequest` -
https://github.com/arhanashik/PSTUian-web-backend/blob/main/app/Http/Requests/FormRequest.php


You can check DonationRequest class.

ManiruzzamanAkash commented on app/Models/Frontend/AccountsOption.php:
```diff
@@ -0,0 +1,15 @@
+    use HasFactory;
+    protected $fillable = [
+        'donation_option',
+        'deleted'
```
Deleted should not be passed as `fillable` property.

ManiruzzamanAkash commented on app/Models/Frontend/AccountsOption.php:
```diff
@@ -0,0 +1,15 @@
+
+class AccountsOption extends Model
+{
+    use HasFactory;
+    protected $fillable = [
```
Please add a line break after
```php
use HasFactory;
```
this line.

ManiruzzamanAkash commented on app/Repositories/AccountOptionRepository.php:
```diff
@@ -0,0 +1,18 @@
+namespace App\Repositories;
+
+use App\Models\Frontend\AccountsOption;
+
+
```
One extra line break, should be only one.

ManiruzzamanAkash commented on app/Repositories/AccountOptionRepository.php:
```diff
@@ -0,0 +1,18 @@
+
+class AccountOptionRepository extends CrudRepository
+{
+    /**
```
I don't thing we need commenting here.

ManiruzzamanAkash commented on app/Services/AccountOptionService.php:
```diff
@@ -0,0 +1,44 @@
+    public function __construct(
+        private readonly AccountOptionRepository $accountOptionRepository
+    ) {}
+
+
+    
```
Keep only one line break. There is three breaks.

ManiruzzamanAkash commented on app/Services/AccountOptionService.php:
```diff
@@ -0,0 +1,44 @@
+    {
+        return $this->accountOptionRepository->create($data);
+    }
+
+
+
```
Three line break here too.

ManiruzzamanAkash commented on app/Services/AccountOptionService.php:
```diff
@@ -0,0 +1,44 @@
+     */
+    public function getAccountOptions(): array
+    {
+        // Delegate to the repository to fetch all account options
```
Remove the comment please from every methods.

ManiruzzamanAkash replied:
Please check Donation Service class.

ManiruzzamanAkash commented on app/Services/AccountOptionService.php:
```diff
@@ -0,0 +1,44 @@
+
+
+    
+    /**
+     * Create a new account option using the provided data.
+     *
+     * @param array $data The data to create the account option.
+     * @return AccountsOption The created account option instance.
+     */
```
I think, we don't need these comments, as this are pretty much and much clear without commenting.

ManiruzzamanAkash commented on database/seeders/AccountOptionSeeder.php:
```diff
@@ -0,0 +1,23 @@
+class AccountOptionSeeder extends Seeder
+{
+    public function __construct(private readonly AccountOptionService $accountOptionService)
+    {
+    }
+    /**
```
Add extra line space after constructor.

ManiruzzamanAkash commented on app/Services/AccountOptionService.php:
```diff
@@ -0,0 +1,44 @@
+        // Delegate to the repository to fetch all account options
+        return $this->accountOptionRepository->all();
+    }
+
+    
+
```
Extra 3 line breaks. remove this.

ManiruzzamanAkash reviewed (commented):
@engr-shishir 
Great Code. Love it. Just a couple of comments. As this is first PR, it is very common to get such reviews. Hope this will be reduced in future PR.

Happy coding, Bro..

engrshishir committed:
upgrade to url friendly route

engrshishir committed:
Code cleanup: removed comments, spaces, and line breaks for readability

developer-friendly code, Like removing extra comments, spaces, and linebreaks. And add the necessary linebreak too

ManiruzzamanAkash committed:
Moved Model from Frontend folder to Models folder

ManiruzzamanAkash committed:
Add required validator in AccountOption request

ManiruzzamanAkash committed:
Fixed phpcs and refactor

ManiruzzamanAkash committed:
Fixed backend route path
